### ISSUE TITLE
Catch "relationship/type" inconsistency for TOSCA 1.3 

### ISSUE THREAD

**Issue by cankarm** [xlab-si/xopera-opera#28]
Opera correctly parses the TOSCA 1.3 version and does work only if:

`        relationship: con_HostedOn_0`

and not if 
```
        relationship:
          type: con_HostedOn_0
```
That's OK. The problem with catching this error and its output. In some cases this raises error with output on different place e.g.:

`<some dir>/<somefilename>.tosca:21:13: [ReferenceWrapper] Expected string input`

The actual error was somewhere else. Is it possible to catch this error before in some pre-check? Clearing white-spaces&grep already solves the solution, but more sophisticated and integrated solution would be appreciated ;).

**Comment by anzoman** [comment]
@cankarm thanks for the enhancement. I will try to update catching errors and add some pre-checks for this type of error.

**Comment by anzoman** [comment]
So here we are talking about an extended relationship TOSCA syntax for the node type requirements. This would probably look like in the example below (I found this extended syntax in section 3.8.2.2.3 of the [TOSCA simple YAML profile](https://docs.oasis-open.org/tosca/TOSCA-Simple-Profile-YAML/v1.3/TOSCA-Simple-Profile-YAML-v1.3.html) ).

```
<requirement_name>:
  relationship:
    type: <relationship_template_name> | <relationship_type_name>
    properties:
      <property_assignments>
    interfaces:
      <interface_assignments>
```

I think that currently opera expects a string reference to the relationship type which needs to be defined separately and used like in the example below.

```
  my.nodes.Nginx.Site:
    derived_from: tosca.nodes.SoftwareComponent
    requirements:
      - host:
          capability: tosca.capabilities.Compute
          relationship: my.relationships.NginxSiteHosting

relationship_types:
  my.relationships.NginxSiteHosting:
    derived_from: tosca.relationships.HostedOn
    valid_target_types: [my.nodes.Nginx]
    interfaces:
      Configure:
        add_source:
          implementation:
            primary: playbooks/nginx/reload.yml
```

If we add the extended syntax this example would probably look like:

```
  my.nodes.Nginx.Site:
    derived_from: tosca.nodes.SoftwareComponent
    requirements:
      - host:
          capability: tosca.capabilities.Compute
          relationship:
          type: my.relationships.NginxSiteHosting
          derived_from: tosca.relationships.HostedOn
          valid_target_types: [my.nodes.Nginx]
          interfaces:
            Configure:
              add_source:
                implementation:
                  primary: playbooks/nginx/reload.yml
```

@tadeboro perhaps you could give us some additional info. I had a look at the TOSCA parser for xOpera and I figured out that this extended syntax is probably not supported at the moment. Right? Do you think that it is wise to implement this for opera?

**Comment by tadeboro** [comment]
This is the can of worms I did not implement because it introduces quite a lot of potential complexity into the parser as well as service templates. The main two reasons why I did not go and implement this are:

 1. I wanted to avoid any runtime selection of relationship templates. This kind of functionality has no place in a document that should deterministically describe a deploy.
 2. Using extended syntax for requirement assignments mixes up type and template definitions in one big pile of problems. Allowing end-users to monkey patch the templates is just calling for troubles.

The examples that @anzoman provided are not strictly related to requirement assignments (3.8.2) since they are demonstrating a mix od requirement definitions (3.7.3) and requirement assignments usage, but they are right in a general sense: that this functionality is not supported and indeed demonstrate how monkey patching should be replaced by explicitly defining some types.

Roughly the same thing can be said about the templates. But the mess one is allowed to make there is even greater:

```
topology_template:
  node_templates:
    nginx:
      type: nginx.Server
      requirements:
        - host:
            node: ...
            relationship:
              type: my.relationships.NginxSiteHosting
              properties:
                a: b
              interfaces:
                Configure:
                  add_source:
                    implementation:
                      primary: playbooks/nginx/reload.yml
```

This should instead be writen as:

```
relationship_types:
  my.rel.Type:
    derived_from: my.relationships.NginxSiteHosting
    interfaces:
      Configure:
        add_source:
          implementation:
            primary: playbooks/nginx/reload.yml

topology_template:
  relationship_templates:
    my_rel_template:
      type: my.rel.Type
      properties:
        a: b

  node_templates:
    nginx:
      type: nginx.Server
      requirements:
        - host:
            node: ...
            relationship: my_rel_template
```

I do not mind if anyone is interested in implementing this, but I am pretty sure that this will be an endless source of problems in the long run. So I completely agree with @cankarm here: adding more meaningful error messages is the way to go here. This will make our lives much easier and guide end-users towards more maintainable templates.

**Comment by anzoman** [comment]
@tadeboro thanks a lot for the detailed explanation. I'm aware that providing the support for this kind of extended syntax can cause further damage. And your first example with the extended syntax is the warning for the potential headache. So, I totally agree wiith you that this could bring a lot of trouble and would result in complicating the life of the users in ways they can't expect. One look at the first example is enough to be confused. One can clearly see that the second example provides a much nicer description for the topology of the cloud application and as you put it monkey pathching could also be a problem. 

And for sure users of xOpera deserve nicer error messages. I will take a closer look at the code where errors are thrown and will try to add those meaningful messages that will help in resolving issues regarding relationship requirements within the TOSCA templates.

**Comment by anzoman** [comment]
@cankarm this was fixed in #29 so this will be completely resolved in the future with the new opera release.

**Comment by cankarm** [comment]
👍 
Thanks, @anzoman!

**Comment by anzoman** [comment]
@cankarm I'm closing this issue since this was already resolved within the last two xOpera versions (0.5.6 and 0.5.7).
